common/persistence/historyStore.go

```diff
@@ -408,83 +402,100 @@ func (m *historyV2ManagerImpl) readRawHistoryBranch(
-		events, err := m.historySerializer.DeserializeBatchEvents(batch)
-		if err != nil {
-			return nil, nil, nil, 0, 0, err
-		}
-		if len(events) == 0 {
-			logger.Error("Empty events in a batch")
-			return nil, nil, nil, 0, 0, serviceerror.NewInternal(fmt.Sprintf("corrupted history event batch, empty events"))
-		}
-
-		firstEvent := events[0]           // first
-		eventCount := len(events)         // length
-		lastEvent := events[eventCount-1] // last
-
-		if firstEvent.GetVersion() != lastEvent.GetVersion() || firstEvent.GetEventId()+int64(eventCount-1) != lastEvent.GetEventId() {
-			// in a single batch, version should be the same, and ID should be continous
-			logger.Error("Corrupted event batch",
-				tag.FirstEventVersion(firstEvent.GetVersion()), tag.WorkflowFirstEventID(firstEvent.GetEventId()),
-				tag.LastEventVersion(lastEvent.GetVersion()), tag.WorkflowNextEventID(lastEvent.GetEventId()),
-				tag.Counter(eventCount))
-			return nil, nil, nil, 0, 0, serviceerror.NewInternal(fmt.Sprintf("corrupted history event batch, wrong version and IDs"))
-		}
-
-		if firstEvent.GetVersion() < token.LastEventVersion {
-			// version decrease means the this batch are all stale events, we should skip
-			logger.Info("Stale event batch with smaller version", tag.FirstEventVersion(firstEvent.GetVersion()), tag.TokenLastEventVersion(token.LastEventVersion))
-			continue
-		}
-		if firstEvent.GetEventId() <= token.LastEventID {
-			// we could see it because first batch of next page has a smaller txn_id
-			logger.Info("Stale event batch with eventID", tag.WorkflowFirstEventID(firstEvent.GetEventId()), tag.TokenLastEventID(token.LastEventID))
-			continue
-		}
-		if firstEvent.GetEventId() != token.LastEventID+1 {
-			// We assume application layer want to read from MinEventID(inclusive)
-			// However, for getting history from remote cluster, there is scenario that we have to read from middle without knowing the firstEventID.
-			// In that case we don't validate history continuousness for the first page
-			// TODO: in this case, some events returned can be invalid(stale). application layer need to make sure it won't make any problems to XDC
-			if defaultLastEventID == 0 || token.LastEventID != defaultLastEventID {
-				logger.Error("Corrupted non-continuous event batch",
-					tag.FirstEventVersion(firstEvent.GetVersion()), tag.WorkflowFirstEventID(firstEvent.GetEventId()),
-					tag.LastEventVersion(lastEvent.GetVersion()), tag.WorkflowNextEventID(lastEvent.GetEventId()),
-					tag.TokenLastEventVersion(token.LastEventVersion), tag.TokenLastEventID(token.LastEventID),
-					tag.Counter(eventCount))
-				return nil, nil, nil, 0, 0, serviceerror.NewInternal(fmt.Sprintf("corrupted history event batch, eventID is not continuous"))
+		var events []*historypb.HistoryEvent
+		events, batchErr = m.validateEventBatch(batch, token, request.MinEventID-1, logger)
+		if events == nil {
+			if batchErr != nil {
+				break
+			} else {
+				continue
```

Comment:
Why are we special casing empty event case and continue to load more batches?
I think we should treat empty event exactly in the same manner as other errors and not proceed further.  This might change the behavior for existing semantics.  Imagine the following situation:

1. Read 10 batches successfully
2. Read empty events batch
3. Read 11th batch successfully

Previously the following sequence of events will result into this API returning an error back to the caller on step 2.  Now it will return a successful response.

Reply:
This just maintains previous behavior where we had a few cases for ignoring batches.
Specifically:
```
if firstEvent.GetVersion() < token.LastEventVersion {
			// version decrease means the this batch are all stale events, we should skip
			logger.Info("Stale event batch with smaller version", tag.FirstEventVersion(firstEvent.GetVersion()), tag.TokenLastEventVersion(token.LastEventVersion))
			continue
		}
		if firstEvent.GetEventId() <= token.LastEventID {
			// we could see it because first batch of next page has a smaller txn_id
			logger.Info("Stale event batch with eventID", tag.WorkflowFirstEventID(firstEvent.GetEventId()), tag.TokenLastEventID(token.LastEventID))
			continue
		}
```

Reply:
Here is the case which will now behave differently.

Code Before your changes
```
events, err := m.historySerializer.DeserializeBatchEvents(batch)
		if err != nil {
			return nil, nil, nil, 0, 0, err
		}
		if len(events) == 0 {
			logger.Error("Empty events in a batch")
			return nil, nil, nil, 0, 0, serviceerror.NewInternal(fmt.Sprintf("corrupted history event batch, empty events"))
		}
```

Based on the previous logic function would return an error immediately when it runs into an empty batch.  Now with the new refactored implementation it will skip the empty batch and continue processing further events which will result into a successful response in the scenario I provided above.  

This function is already quite tricky and I don't believe the refactored implementation buys us anything.  I would rather leave the implementation as is.  Considering we already moved the validation part to caller, all we need to do is return the events even in the case of failure and let the caller decide what to do with partial payload.

This is very core functionality and any slight change in semantics might result into very tricky bugs.  I would not try to refactor this logic without understanding impact on all callers of this function.

Reply:
We can revert the refactoring, however for the sake of correctness let's try to complete this argument here.
If the batch is empty, then validation would return (nil, error) (line 455) which in turn will result in breaking the loop (line 423) and returning an error to the caller (line 498). So the only difference would be that previous code would return nil, nil, ..., err but now we'll return all previously collected events, which is exactly what we want here.